apps/consulting/posts/review-functorch-contribution.md

```diff
@@ -0,0 +1,381 @@
+produce an FX graph once again. However, this time, it represents these operations
+at a lower level using `aten` operators. Additionally, `aot_autograd` decomposes
+composite operations into the primitive operations. For instance, a composite
+operation like `torch.square` is traced down to `torch.pow(x, 2)`.
+
+Moreover, `aot_autograd` also manages the creation of the backward graph when
+requested. The name "ahead of time autograd" or `aot_autograd` is derived from
+this ability to handle autograd operations in advance. Below, you can see the
+graph generated by `aot_autograd` for the above program.
+
+```python
+def forward(self, arg0_1: f32[3, 3]):
+    # File: test/test_scratch.py:334, code: return torch.sin(x) + torch.square(x)
+    sin: f32[3, 3] = torch.ops.aten.sin.default(arg0_1)
+    pow_1: f32[3, 3] = torch.ops.aten.pow.Tensor_Scalar(arg0_1, 2);  arg0_1 = None
+    add: f32[3, 3] = torch.ops.aten.add.Tensor(sin, pow_1);  sin = pow_1 = None
+    return (add,)
+
+```
+
+inductor: As discussed above, it is inductor's job to apply optimisations and
+generate specialised code. In this case, it has fused `sin` and `square` to run
+within the same `for`-loop. This allows the generated program to do more compute
+per read/write effectively improving the memory bandwith utilization.
+
+```python
+from ctypes import c_void_p, c_long
+import torch
+import math
+import random
+import os
+import tempfile
+from math import inf, nan
+from torch._inductor.hooks import run_intermediate_hooks
+from torch._inductor.utils import maybe_profile
+
+from torch import empty_strided, device
+from torch._inductor.codecache import AsyncCompile
+from torch._inductor.select_algorithm import extern_kernels
+
+aten = torch.ops.aten
+assert_size_stride = torch._C._dynamo.guards.assert_size_stride
+reinterpret_tensor = torch.ops.inductor._reinterpret_tensor
+async_compile = AsyncCompile()
+
+
+cpp_fused_add_cos_sin_0 = async_compile.cpp('''
+#include "/tmp/torchinductor_kshiteej/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
+extern "C" void kernel(const float* in_ptr0,
+                       float* out_ptr0)
+{
+    {
+        for(long i0=static_cast<long>(0L); i0<static_cast<long>(8L); i0+=static_cast<long>(8L))
+        {
+            auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + static_cast<long>(i0));
+            auto tmp1 = tmp0.sin();
+            auto tmp2 = tmp0.cos();
+            auto tmp3 = tmp1 + tmp2;
+            tmp3.store(out_ptr0 + static_cast<long>(i0));
+        }
+        #pragma omp simd simdlen(4) 
+        for(long i0=static_cast<long>(8L); i0<static_cast<long>(9L); i0+=static_cast<long>(1L))
+        {
+            auto tmp0 = in_ptr0[static_cast<long>(i0)];
+            auto tmp1 = std::sin(tmp0);
+            auto tmp2 = std::cos(tmp0);
+            auto tmp3 = tmp1 + tmp2;
+            out_ptr0[static_cast<long>(i0)] = tmp3;
+        }
+    }
+}
```

Comment:
Perhaps just put this bit, and simplify the formatting a bit so that it's more readable?